
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Single-Cell Image Analysis Package">
      
      
        <meta name="author" content="Ajit Johnson Nirmal">
      
      
      
        <link rel="prev" href="../1-scimap-tutorial-getting-started/">
      
      
        <link rel="next" href="../3-Cell_Type_calling_and_adding_ROIs/">
      
      
        
      
      
      <link rel="icon" href="../../../assets/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>Cell-Phenotyping using Scimap - scimap</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-22X9HCL4YP"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-22X9HCL4YP",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-22X9HCL4YP",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cell-phenotyping-using-scimap" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="scimap" class="md-header__button md-logo" aria-label="scimap" data-md-component="logo">
      
  <img src="../../../assets/favicon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            scimap
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Cell-Phenotyping using Scimap
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="white" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/ajitjohnson/scimap" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../Getting%20Started/" class="md-tabs__link">
        
  
  
    
  
  Install

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../Tools%20Shortcut/" class="md-tabs__link">
        
  
  
    
  
  Index

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../Functions/pp/mcmicro_to_scimap/" class="md-tabs__link">
          
  
  
  Functions

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../md/spatial_biology_scimap/" class="md-tabs__link">
          
  
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../contribute/" class="md-tabs__link">
        
  
  
    
  
  Contribute

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="scimap" class="md-nav__button md-logo" aria-label="scimap" data-md-component="logo">
      
  <img src="../../../assets/favicon.png" alt="logo">

    </a>
    scimap
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/ajitjohnson/scimap" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Getting%20Started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Install
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Tools%20Shortcut/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Index
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Functions
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Functions
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Pre Processing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Pre Processing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pp/mcmicro_to_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    mcmicro_to_scimap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pp/log1p/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    log1p
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pp/rescale/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    rescale
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pp/combat/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    combat
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Tools
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Tools
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/phenotype_cells/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    phenotype_cells
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/cluster/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    cluster
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/umap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    umap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/foldchange/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    foldchange
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_distance/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_distance
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_interaction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_interaction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_count/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_count
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_lda/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_lda
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_expression/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_cluster/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_cluster
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_pscore/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_pscore
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_aggregate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_aggregate
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/tl/spatial_similarity_search/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_similarity_search
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Plotting Functions
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Plotting Functions
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/image_viewer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    image_viewer
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/addROI_image/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    addROI_image
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/gate_finder/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    gate_finder
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/napariGater/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    napariGater
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/heatmap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    heatmap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/markerCorrelation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    markerCorrelation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/groupCorrelation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    groupCorrelation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/distPlot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    distPlot
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/densityPlot2D/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    densityPlot2D
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/cluster_plots/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    cluster_plots
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/umap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    umap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/foldchange/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    foldchange
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/spatial_scatterPlot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_scatterPlot
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/spatial_distance/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_distance
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/spatial_interaction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_interaction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/spatialInteractionNetwork/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatialInteractionNetwork
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/spatial_pscore/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    spatial_pscore
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/stacked_barplot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    stacked_barplot
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/pie/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    pie
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/pl/voronoi/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    voronoi
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Helper Functions
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Helper Functions
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/classify/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    classify
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/rename/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    rename
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/addROI_omero/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    addROI_omero
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/dropFeatures/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    dropFeatures
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/merge_adata_obs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    merge_adata_obs
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/animate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    animate
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Functions/hl/scimap_to_csv/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    scimap_to_csv
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Tutorials
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Tutorials
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/spatial_biology_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Index
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Setup
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Setup
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/install_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Install Scimap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/demo_data_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo Data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/anndata_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Prepare Demo
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Cell Phenotyping
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Cell Phenotyping
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/clustering_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Unsupervised clustering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/scimap_phenotyping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Prior knowledge driven
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/add_rois_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Add ROIs
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/cell_proportion_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Explore cell types
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Spatial Analysis
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Spatial Analysis
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/spatial_analysis_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Distance measurement
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/cell_interaction_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Interaction analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/proximity_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Proximity scores
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/spatial_search_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Search patterns
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          <label class="md-nav__link" for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Cellular Neighbourhoods
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Cellular Neighbourhoods
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/spatial_lda_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Spatial LDA
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/spatial_lag_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Spatial Lag
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_6" >
        
          
          <label class="md-nav__link" for="__nav_5_6" id="__nav_5_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Helper
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Helper
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/export_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Export data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/animate_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Animate
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/scimap_helper_functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Other helpers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../md/scimap_video/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Videos
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_8" checked>
        
          
          <label class="md-nav__link" for="__nav_5_8" id="__nav_5_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Archive
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_8_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Archive
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../1-scimap-tutorial-getting-started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Getting Started with Scimap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Cell-Phenotyping using Scimap
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Cell-Phenotyping using Scimap
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial-material" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial material
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-video" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial video
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clustering-and-data-exploration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Clustering and data exploration
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clustering and data exploration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#finding-marker-genes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Finding marker genes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#probability-distribution-based-phenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      
        Probability distribution based phenotyping
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Probability distribution based phenotyping">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#in-order-to-run-the-method-you-need-2-things" class="md-nav__link">
    <span class="md-ellipsis">
      
        In order to run the method, you need 2 things
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#define-manual-gates-to-rescale-data-before-running-the-phenotyping-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      
        Define manual gates to rescale data before running the phenotyping algorithm
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-1-identify-the-gates-using-smplgate_finder" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1: Identify the gates using sm.pl.gate_finder
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-rescale-the-data-based-on-the-identified-gates-using-smpprescale" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2: Rescale the data based on the identified gates using sm.pp.rescale
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#note-for-v0220" class="md-nav__link">
    <span class="md-ellipsis">
      
        Note (for &gt;=v.0.22.0)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-run-the-phenotyping-algorithm-on-the-rescaled-data-using-smtlphenotype" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 3: Run the phenotyping algorithm on the rescaled data using sm.tl.phenotype
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3: Run the phenotyping algorithm on the rescaled data using sm.tl.phenotype">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#heatmap-and-umap-of-the-probability-based-phenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      
        Heatmap and UMAP of the probability based phenotyping
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../3-Cell_Type_calling_and_adding_ROIs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Cell-Phenotyping and adding ROIs
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../4-CellType_Proportion_Exploration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    CellType Proportion Exploration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../5-Simple_Spatial_Analysis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Simple Spatial Analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../6_animate_with_scimap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Animate with scimap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../contribute/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Contribute
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial-material" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial material
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-video" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial video
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clustering-and-data-exploration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Clustering and data exploration
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clustering and data exploration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#finding-marker-genes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Finding marker genes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#probability-distribution-based-phenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      
        Probability distribution based phenotyping
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Probability distribution based phenotyping">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#in-order-to-run-the-method-you-need-2-things" class="md-nav__link">
    <span class="md-ellipsis">
      
        In order to run the method, you need 2 things
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#define-manual-gates-to-rescale-data-before-running-the-phenotyping-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      
        Define manual gates to rescale data before running the phenotyping algorithm
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-1-identify-the-gates-using-smplgate_finder" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1: Identify the gates using sm.pl.gate_finder
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-rescale-the-data-based-on-the-identified-gates-using-smpprescale" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2: Rescale the data based on the identified gates using sm.pp.rescale
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#note-for-v0220" class="md-nav__link">
    <span class="md-ellipsis">
      
        Note (for &gt;=v.0.22.0)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-run-the-phenotyping-algorithm-on-the-rescaled-data-using-smtlphenotype" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 3: Run the phenotyping algorithm on the rescaled data using sm.tl.phenotype
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3: Run the phenotyping algorithm on the rescaled data using sm.tl.phenotype">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#heatmap-and-umap-of-the-probability-based-phenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      
        Heatmap and UMAP of the probability based phenotyping
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="cell-phenotyping-using-scimap">Cell-Phenotyping using Scimap</h1>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Created on Fri Jun 28 18:10:06 2020</span>
<span class="sd">@author: Ajit Johnson Nirmal</span>
<span class="sd">Scimap Cell Phenotyping Tutorial</span>
<span class="sd">&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<div class="codehilite"><pre><span></span><code><span class="s1">&#39;\nCreated on Fri Jun 28 18:10:06 2020\n@author: Ajit Johnson Nirmal\nScimap Cell Phenotyping Tutorial\n&#39;</span>
</code></pre></div>

<h2 id="tutorial-material">Tutorial material</h2>
<p>You can download the material for this tutorial from the following <a href="https://www.dropbox.com/s/rra13zir52o9hio/getting_started%20and%20phenotyping.zip?dl=0">link:</a><br />
The presentation files are available <a href="https://github.com/ajitjohnson/Jupyter-Notebooks/blob/master/tutorials/scimap_tutorial/getting_started%20and%20phenotyping/scimap_tutorial.pdf">here:</a></p>
<h2 id="tutorial-video">Tutorial video</h2>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">HTML</span>
<span class="n">HTML</span><span class="p">(</span><span class="s1">&#39;&lt;iframe width=&quot;450&quot; height=&quot;250&quot; src=&quot;https://www.youtube.com/embed/knh5elRksUk&quot; frameborder=&quot;0&quot; allow=&quot;accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture&quot; allowfullscreen&gt;&lt;/iframe&gt;&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<iframe width="450" height="250" src="https://www.youtube.com/embed/knh5elRksUk" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Load necessary libraries</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span><span class="p">;</span> <span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">color_codes</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Import Scimap</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scimap</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Set the working directory</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span> <span class="p">(</span><span class="s2">&quot;/Users/aj/Desktop/scimap_tutorial/&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Load data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;tutorial_data.h5ad&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<h2 id="clustering-and-data-exploration">Clustering and data exploration</h2>
<p>You could use clustering and marker expression analysis within clusters to assign cell types similar to what is carried out with single-cell sequencing data.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="c1"># Computing the neighborhood graph</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span> <span class="c1"># Build a UMAP to visualize the neighbourhood graph</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">,</span> <span class="s1">&#39;CD20&#39;</span><span class="p">,</span> <span class="s1">&#39;CD163&#39;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span> <span class="s1">&#39;vlag&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span> <span class="c1"># Plot the UMAP</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_13_0.png" /></p>
<p>We can already begin to spot issues with carrying out this mode of phenotyping approach. As you can see there is an area of co-expression of CD3D and CD20, which is likely because of segmentation errors. Additionally the boundaries are not distinct between cell-types and it is highly likely that errors will be introduced due to this reason. </p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># Clustering the neighborhood graph</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3D&#39;</span><span class="p">,</span> <span class="s1">&#39;CD20&#39;</span><span class="p">],</span><span class="n">cmap</span><span class="o">=</span> <span class="s1">&#39;vlag&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1"># View the clustering</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_16_0.png" /></p>
<h3 id="finding-marker-genes">Finding marker genes</h3>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_18_0.png" /></p>
<p>From the above plots, it is likely that clusters 1, 2 and 7 could be combined to form a T cell cluster. However, as mentioned earlier the boundaries are not clear and it only get increasingly complex as one would want to perform deeper phenotyping such as CD4 helper T cells, CD8 T cells, regulatory T cells and so on. </p>
<p>Additionally, marker analsyis suggests that CD30 is most expressed in cluster 8. If you look at the actual image, you will realize that CD30 is not expressed by any cell in this image and the analysis is picking up high background fluorescence. </p>
<h2 id="probability-distribution-based-phenotyping">Probability distribution based phenotyping</h2>
<p>This approach is more labor intensive, however is significantly more sensitive and much more scalable than clustering based approaches. Takes less than 5 mins to run over a million cells once the gates are identified.</p>
<h4 id="in-order-to-run-the-method-you-need-2-things">In order to run the method, you need 2 things</h4>
<ul>
<li>a gating workflow strategy <code>.csv file</code></li>
<li>manual gates <code>.csv file</code>. If manual gates are not provided, the algorithm will attempt to rescale the data by fitting two gaussians on the data. However, it is adviced to perform manual gating as I have found it to be more sensitive.</li>
</ul>
<p><strong>The algorithm involves three steps:</strong>
1. Identify the gates using <code>sm.pl.gate_finder</code>
2. Rescale the data based on the identified gates using <code>sm.pp.rescale</code>
3. Run the phenotyping algorithm on the rescaled data using <code>sm.tl.phenotype</code></p>
<h4 id="define-manual-gates-to-rescale-data-before-running-the-phenotyping-algorithm">Define manual gates to rescale data before running the phenotyping algorithm</h4>
<p>Instantiating the Qt GUI can take a few seconds and if you create the Viewer before it is finished, the kernel will die and the viewer will not launch. For this reason the %gui qt magic command should always be run in a separate cell from creating the viewer</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="o">%</span><span class="n">gui</span> <span class="n">qt</span>
</code></pre></div></td></tr></table></div>
<h3 id="step-1-identify-the-gates-using-smplgate_finder">Step 1: Identify the gates using <code>sm.pl.gate_finder</code></h3>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">image_path</span> <span class="o">=</span> <span class="s1">&#39;/Users/aj/Desktop/scimap_tutorial/reactive_core.tif&#39;</span>
<span class="n">marker_of_interest</span> <span class="o">=</span> <span class="s1">&#39;CD45&#39;</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gate_finder</span> <span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">marker_of_interest</span><span class="p">,</span> 
                   <span class="n">from_gate</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">to_gate</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span> <span class="n">increment</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> 
                   <span class="n">markers</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ASMA&#39;</span><span class="p">,</span><span class="s1">&#39;DNA11&#39;</span><span class="p">,</span> <span class="s1">&#39;CD20&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3D&#39;</span><span class="p">],</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<h3 id="step-2-rescale-the-data-based-on-the-identified-gates-using-smpprescale">Step 2: Rescale the data based on the identified gates using <code>sm.pp.rescale</code></h3>
<p>Note: Below we are passing a <code>manual_gates.csv</code> into the <code>gate</code> parameter.
This contatins gates that were visually determined using the <code>sm.pl.gate_finder</code>
function. For the markers included in the <code>manual_gates.csv</code> file, 
the function will scale the data such that cells with expression greater than the gate 
will be considered as positive for that marker and cells
with expression below the gate is considered negative. <br></p>
<p>For markers that are not included in the <code>manual_gates.csv</code> file, the function
will automatically try to determine a gate by running a gaussian mixture model
algorithm on the data. <br></p>
<h3 id="note-for-v0220">Note (for &gt;=v.0.22.0)</h3>
<p>Please note that passing manual gates for multiple images has been introduced in <code>scimap &gt;=v.0.22.0</code></p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Load the manual gates and rescale the data based on the gates</span>
<span class="n">manual_gate</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;manual_gates.csv&#39;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">rescale</span> <span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gate</span><span class="o">=</span><span class="n">manual_gate</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<div class="codehilite"><pre><span></span><code><span class="n">Scaling</span><span class="w"> </span><span class="nc">Image</span><span class="w"> </span><span class="o">[</span><span class="n">mcmicro_output</span><span class="o">]</span>
<span class="n">Categories</span><span class="w"> </span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="k">object</span><span class="p">)</span><span class="err">:</span><span class="w"> </span><span class="o">[</span><span class="n">mcmicro_output</span><span class="o">]</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD10</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD2</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD30</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD43</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD5</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD57</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">CD7</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">KI67</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">MHCI</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">PDL1</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">PS6</span>
<span class="n">Finding</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">optimal</span><span class="w"> </span><span class="n">gate</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">PSTAT3</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">ASMA</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD163</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD206</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD68</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD20</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD21</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD3D</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD45</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD56</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD8A</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">FOXP3</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD11B</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD11C</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD15</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD4</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">PD1</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">HLADR</span>
<span class="n">Scaling</span><span class="w"> </span><span class="n">CD25</span>
</code></pre></div>

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># View the scaled data (note that the log data is replaced with scaled data)</span>
<span class="c1"># If you ever want the log data back you will need to run-  np.log1p(adata.raw.X)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span> 
</code></pre></div></td></tr></table></div>
<div class="codehilite"><pre><span></span><code>array([[0.17841106, 0.45723783, 0.49234127, ..., 0.15973007, 0.1665647 ,
        0.20024123],
       [0.155838  , 0.21377199, 0.34924023, ..., 0.1522421 , 0.0885678 ,
        0.15338667],
       [0.29090098, 0.37456273, 0.50752728, ..., 0.21580293, 0.17788475,
        0.18778977],
       ...,
       [0.33621626, 0.70161347, 0.70359459, ..., 0.26182348, 0.60810172,
        0.22068843],
       [0.15324935, 0.51044092, 0.60877724, ..., 0.11845023, 0.26558105,
        0.08391592],
       [0.18923565, 0.431478  , 0.58019522, ..., 0.09423545, 0.24047052,
        0.12733514]])
</code></pre></div>

<h3 id="step-3-run-the-phenotyping-algorithm-on-the-rescaled-data-using-smtlphenotype">Step 3: Run the phenotyping algorithm on the rescaled data using <code>sm.tl.phenotype</code></h3>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Load the gating workflow</span>
<span class="n">phenotype</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;phenotype_workflow.csv&#39;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">phenotype_cells</span> <span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">phenotype</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;phenotype&quot;</span><span class="p">)</span> 
</code></pre></div></td></tr></table></div>
<div class="codehilite"><pre><span></span><code><span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Other</span><span class="w"> </span><span class="nx">Immune</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">ASMA</span><span class="o">+</span><span class="w"> </span><span class="nx">cells</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">Other</span><span class="w"> </span><span class="nx">Immune</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">B</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Lineage</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">NK</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Granulocytes</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Lineage</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">B</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">NK</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Granulocytes</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">CD68</span><span class="o">+</span><span class="w"> </span><span class="nx">Macrophages</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">M2</span><span class="w"> </span><span class="nx">Macrophages</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Follicular</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">CD4</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">CD8</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">CD4</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Regulatory</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">Follicular</span><span class="w"> </span><span class="nx">Helper</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">CD8</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">PD1</span><span class="o">+</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span>
<span class="o">--</span><span class="w"> </span><span class="nx">Subsetting</span><span class="w"> </span><span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Phenotyping</span><span class="w"> </span><span class="nx">CD25</span><span class="o">+</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span>
<span class="nx">Consolidating</span><span class="w"> </span><span class="nx">the</span><span class="w"> </span><span class="nx">phenotypes</span><span class="w"> </span><span class="nx">across</span><span class="w"> </span><span class="nx">all</span><span class="w"> </span><span class="nx">groups</span>
</code></pre></div>

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Summary of the phenotyping</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
<div class="codehilite"><pre><span></span><code><span class="nx">B</span><span class="w"> </span><span class="nx">cells</span><span class="w">                       </span><span class="mi">2037</span>
<span class="nx">CD4</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">                    </span><span class="mi">502</span>
<span class="nx">ASMA</span><span class="o">+</span><span class="w"> </span><span class="nx">cells</span><span class="w">                    </span><span class="mi">420</span>
<span class="nx">Regulatory</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">             </span><span class="mi">418</span>
<span class="nx">CD8</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">                    </span><span class="mi">322</span>
<span class="nx">Follicular</span><span class="w"> </span><span class="nx">Helper</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">      </span><span class="mi">282</span>
<span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">                        </span><span class="mi">146</span>
<span class="nx">Unknown</span><span class="w">                        </span><span class="mi">140</span>
<span class="nx">Other</span><span class="w"> </span><span class="nx">Immune</span><span class="w"> </span><span class="nx">cells</span><span class="w">             </span><span class="mi">137</span>
<span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span><span class="w">        </span><span class="mi">124</span>
<span class="nx">Follicular</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span><span class="w">      </span><span class="mi">87</span>
<span class="nx">Myeloid</span><span class="w"> </span><span class="nx">Lineage</span><span class="w">                 </span><span class="mi">80</span>
<span class="nx">PD1</span><span class="o">+</span><span class="w"> </span><span class="nx">T</span><span class="w"> </span><span class="nx">cells</span><span class="w">                    </span><span class="mi">63</span>
<span class="nx">M2</span><span class="w"> </span><span class="nx">Macrophages</span><span class="w">                  </span><span class="mi">55</span>
<span class="nx">Granulocytes</span><span class="w">                     </span><span class="mi">8</span>
<span class="nx">NK</span><span class="w"> </span><span class="nx">cells</span><span class="w">                         </span><span class="mi">3</span>
<span class="nx">CD25</span><span class="o">+</span><span class="w"> </span><span class="nx">Dendritic</span><span class="w"> </span><span class="nx">cells</span><span class="w">            </span><span class="mi">1</span>
<span class="nx">Name</span><span class="p">:</span><span class="w"> </span><span class="nx">phenotype</span><span class="p">,</span><span class="w"> </span><span class="nx">dtype</span><span class="p">:</span><span class="w"> </span><span class="nx">int64</span>
</code></pre></div>

<p>It is likely that <code>CD25+ Dendritic cells, NK cells &amp; Granulocytes</code> are artifacts. You could set <code>pheno_threshold_abs= 10</code> to move these cells into <code>unknown</code> category.</p>
<p><strong>Once the phenotyping is performed, it is adviced to overlay the phenotypes on the image and check if they are correct. If not, alter the <code>phenotyping workflow</code> file or the <code>manual gate</code> to account for the errors.</strong></p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># View phenotypes</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">image_viewer</span> <span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">overlay</span> <span class="o">=</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">point_color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># View Leiden clustering</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">image_viewer</span> <span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">overlay</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">point_color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<h4 id="heatmap-and-umap-of-the-probability-based-phenotyping">Heatmap and UMAP of the probability based phenotyping</h4>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var_names</span><span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">dendrogram</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;vlag&quot;</span><span class="p">,</span> <span class="n">standard_scale</span><span class="o">=</span><span class="s1">&#39;var&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_40_0.png" /></p>
<div class="codehilite"><pre><span></span><code>GridSpec(2, 3, height_ratios=[0, 10.5], width_ratios=[9.6, 0.8, 0.2])
</code></pre></div>

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">])</span> <span class="c1"># View the clustering</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_41_0.png" /></p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">:(</span><span class="mf">11.7</span><span class="p">,</span><span class="mf">8.27</span><span class="p">)})</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">],</span><span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span> <span class="c1"># View the clustering</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_42_0.png" /></p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">,</span> <span class="s1">&#39;PD1&#39;</span><span class="p">,</span> <span class="s1">&#39;CD20&#39;</span><span class="p">],</span><span class="n">cmap</span><span class="o">=</span> <span class="s1">&#39;vlag&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span> <span class="c1"># View the clustering</span>
</code></pre></div></td></tr></table></div>
<p><img alt="png" src="../scimap-tutorial-cell-phenotyping_files/scimap-tutorial-cell-phenotyping_43_0.png" /></p>
<p>As it can be seen from above 3 UMAP's it would have been very difficult to find the Follicular helper T cells by a pure clustering approach. Also, the B cells as can be seen above does not from a nice seperate cluster. These among other illustrate the importance of using the probability based algorithm for deep phenotyping.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Confirm Follicular helper T cells in the image</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">image_viewer</span> <span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> 
                    <span class="n">overlay</span> <span class="o">=</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">overlay_category</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Follicular Helper T cells&#39;</span><span class="p">],</span> 
                    <span class="n">markers</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">,</span><span class="s1">&#39;CD20&#39;</span><span class="p">,</span><span class="s1">&#39;PD1&#39;</span><span class="p">,</span><span class="s1">&#39;CD8A&#39;</span><span class="p">,</span><span class="s1">&#39;CD4&#39;</span><span class="p">,</span><span class="s1">&#39;DNA11&#39;</span><span class="p">],</span>
                    <span class="n">point_color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Save the results</span>
<span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;tutorial_data.h5ad&#39;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<p><strong>This concludes this tutorial</strong></p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 | Nirmal Lab | www.nirmallab.com
    </div>
  
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
    <a href="https://twitter.com/nirmallab" target="_blank" rel="noopener" title="Twitter" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M459.4 151.7c.3 4.5.3 9.1.3 13.6 0 138.7-105.6 298.6-298.6 298.6-59.5 0-114.7-17.2-161.1-47.1 8.4 1 16.6 1.3 25.3 1.3 49.1 0 94.2-16.6 130.3-44.8-46.1-1-84.8-31.2-98.1-72.8 6.5 1 13 1.6 19.8 1.6 9.4 0 18.8-1.3 27.6-3.6-48.1-9.7-84.1-52-84.1-103v-1.3c14 7.8 30.2 12.7 47.4 13.3-28.3-18.8-46.8-51-46.8-87.4 0-19.5 5.2-37.4 14.3-53C87.4 130.8 165 172.4 252.1 176.9c-1.6-7.8-2.6-15.9-2.6-24C249.5 95.1 296.3 48 354.4 48c30.2 0 57.5 12.7 76.7 33.1 23.7-4.5 46.5-13.3 66.6-25.3-7.8 24.4-24.4 44.8-46.1 57.8 21.1-2.3 41.6-8.1 60.4-16.2-14.3 20.8-32.2 39.3-52.6 54.3"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight"], "search": "../../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>